llama-cpp-sys-2/build.rs

```diff
@@ -1,187 +1,41 @@
+use cmake::Config;
 use std::env;
-use std::path::Path;
-use std::path::PathBuf;
+use std::path::{Path, PathBuf};
 
 fn main() {
     println!("cargo:rerun-if-changed=llama.cpp");
 
-    let cublas_enabled = env::var("CARGO_FEATURE_CUBLAS").is_ok();
-
-    let mut ggml_cuda = if cublas_enabled {
-        Some(cc::Build::new())
-    } else {
-        None
-    };
-
     if !Path::new("llama.cpp/ggml.c").exists() {
         panic!("llama.cpp seems to not be populated, try running `git submodule update --init --recursive` to init.")
     }
 
-    let mut ggml = cc::Build::new();
-    let mut llama_cpp = cc::Build::new();
-
-    ggml.cpp(false);
-    llama_cpp.cpp(true);
-
-    // https://github.com/ggerganov/llama.cpp/blob/a836c8f534ab789b02da149fbdaf7735500bff74/Makefile#L364-L368
-    if let Some(ggml_cuda) = &mut ggml_cuda {
-        for lib in [
-            "cuda", "cublas", "culibos", "cudart", "cublasLt", "pthread", "dl", "rt",
-        ] {
-            println!("cargo:rustc-link-lib={}", lib);
-        }
-        if !ggml_cuda.get_compiler().is_like_msvc() {
-            for lib in ["culibos", "pthread", "dl", "rt"] {
-                println!("cargo:rustc-link-lib={}", lib);
-            }
-        }
-
-        println!("cargo:rustc-link-search=native=/usr/local/cuda/lib64");
-
-        if cfg!(target_arch = "aarch64") {
-            ggml_cuda
-                .flag_if_supported("-mfp16-format=ieee")
-                .flag_if_supported("-mno-unaligned-access");
-            ggml.flag_if_supported("-mfp16-format=ieee")
-                .flag_if_supported("-mno-unaligned-access");
-            llama_cpp
-                .flag_if_supported("-mfp16-format=ieee")
-                .flag_if_supported("-mno-unaligned-access");
-            ggml.flag_if_supported("-mfp16-format=ieee")
-                .flag_if_supported("-mno-unaligned-access");
-        }
-
-        ggml_cuda
-            .cuda(true)
-            .flag("-arch=all")
-            .file("llama.cpp/ggml-cuda.cu")
-            .include("llama.cpp");
-
-        if ggml_cuda.get_compiler().is_like_msvc() {
-            ggml_cuda.std("c++14");
-        } else {
-            ggml_cuda.flag("-std=c++11").std("c++11");
-        }
-
-        ggml.define("GGML_USE_CUBLAS", None);
-        ggml_cuda.define("GGML_USE_CUBLAS", None);
-        llama_cpp.define("GGML_USE_CUBLAS", None);
-    }
-
-    for build in [&mut ggml, &mut llama_cpp] {
-        let compiler = build.get_compiler();
-
-        if cfg!(target_arch = "i686") || cfg!(target_arch = "x86_64") {
```

Comment:
Note that by default the CMake build will assume `target-cpu=native` (`LLAMA_NATIVE`), so you'll lose some portability by removing this. I don't mind making a PR to adapt this feature detection to the CMake build after this is merged, if you'd prefer.

Reply:
I can re-add the feature detection (doesn't seem fair to delete your code and expect you to add it back!)

How does the feature detection as we had it differ from targeting native as exists in the cmake file?

What sort of portability do we lose out on?

Reply:
The previous feature detection worked based on the target's features, e.g., if you want to ensure almost any modern CPU can run your software you can use `-Ctarget-cpu=ivybridge` as a compiler flag to enable _just_ `avx` and `sse3`, which any CPU made post 2012 should support. 

With the `LLAMA_NATIVE` flag it depends on the CPU of the one compiling the software, so it will enable all features and ignore your `-Ctarget-cpu=ivybridge` setting. Or in turn, it might not enable AVX2 if it was compiled on an older CPU.

To replicate the old behaviour it's just a case of passing `LLAMA_AVX/2=ON/OFF` etc, to the cmake build instead of the compiler specific flags (and turn off `LLAMA_NATIVE`). Similar to what `llama.cpp` does in their CI.

Reply:
Thanks, that seems worth the effort. You're welcome to make a PR to this branch (won't be merged for a while). 

Otherwise I'll make sure it's done before this gets merged into main.